libp2p/switch.nim

```diff
@@ -291,11 +299,16 @@ proc stop*(s: Switch) {.async.} =
   await allFutures(toSeq(s.connections.values).mapIt(s.cleanupConn(it)))
   await allFutures(s.transports.mapIt(it.close()))
 
-proc subscribeToPeer*(s: Switch, peerInfo: PeerInfo) {.async, gcsafe.} =
+proc subscribeToPeer(s: Switch, peerInfo: PeerInfo) {.async, gcsafe.} =
   ## Subscribe to pub sub peer
-  if s.pubSub.isSome:
-    let conn = await s.dial(peerInfo, s.pubSub.get().codec)
-    await s.pubSub.get().subscribeToPeer(conn)
+  if s.pubSub.isSome and peerInfo.id notin s.dialedPubSubPeers:
+    try:
+      s.dialedPubSubPeers.incl(peerInfo.id)
+      let conn = await s.dial(peerInfo, s.pubSub.get().codec)
+      await s.pubSub.get().subscribeToPeer(conn)
+    except CatchableError as exc:
```

Comment:
so from the perspective of the `subscribeToPeer` caller, this should silently fail?

Reply:
Yes, the reason is that if the other end doesn't support pubsub `dial` will throw, but we want opening pubsub streams to be as non-intrusive as possible. The only thing I can think of making this a little more prominent is turning the trace into a warning.

Reply:
`warn` is for stuff where the user/sysadmin needs to take action - at most notice would be appropriate, or maybe info - however, it seems more appropriate to return a bool here or something - it seems like the kind of thing that a caller would want to know about so they can score the peer, no?

Reply:
Also i think that catching `CatchableError` is a bad practic, you should create your own base exception somewhere in codebase and use it, in such case you will catch all your own exception, but do not catch all others.

Reply:
well, that's something that our code does in general: it doesn't consider errors/exceptions to be part of the api / abstraction, so any kinds of errors are raised anywhere randomly and let free to propagate randomly hoping that someone will catch them - hence, it's impossible to know what kind of exceptions might be caught here, but the code wants to catch them all